Model X-rays Step 1: x-ray base model and test

markbastian opened the pull request:
Fixes #28674

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/28705)
<!-- Reviewable:end -->

markbastian committed:
Doc fixes

markbastian committed:
Adding some comments to undestand x-rays

markbastian committed:
investigation

markbastian committed:
Merge branch 'master' into x-ray-characterization

markbastian committed:
sp

markbastian committed:
Merge branch 'master' into x-ray-characterization

markbastian committed:
Merge branch 'master' into x-ray-characterization

markbastian committed:
Initial commit of x-rays for models

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian commented on src/metabase/automagic_dashboards/core.clj:
```diff
@@ -243,11 +245,15 @@
                              (assoc :entity_type :entity/GenericTable))
     (native-query? card) (-> card (assoc :entity_type :entity/GenericTable))
-    :else                (->> card table-id (db/select-one Table :id))))
+    ;; This is a non-native question with no ancestors (It is a query on a table).
```
This is the only material change to the src branch in this PR. Everything else is cleanup, docs, refactoring for clarification, etc.

markbastian commented on test/metabase/automagic_dashboards/core_test.clj:
```diff
@@ -208,6 +208,110 @@
             (perms/grant-collection-readwrite-permissions! (perms-group/all-users) collection-id)
             (test-automagic-analysis (db/select-one Card :id card-id) 8)))))))
 
```
Step 1: Add some reasonable tests about the actual contents of cards being generated for specific cases. In this PR, the focus is on a model created from a table.

markbastian committed:
linting

dpsutton commented on src/metabase/automagic_dashboards/core.clj:
```diff
@@ -713,10 +719,13 @@
-                             :tables
-                             (m/find-first (comp #{(:table_id field)} u/the-id))
-                             :link)
+                :link (->> context
```
Is it possible to add a helpful error in the case it otherwise blows up? Doesn't improve the functionality, but helps in understanding. An error like "Unable to find Table to link root" or something versus "integer is not seqable" or whatever the blow up is here.

markbastian replied:
Yeah, just need to remember how I reproduced this. It was kind of a corner case. Part of the next several steps is putting together more complicated cases to explicitly unearth and test this stuff.

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Added models as an option in the auto dashboard routes

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Added table "hint" to card to treat as table. (#28806)

Changes are small enough that it's better to combine this with the base branch.

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

ranquild committed:
Add a way to show xrays for questions and models in collection views

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
re-staging my changes

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Tests for more complex model

markbastian commented on src/metabase/api/automagic_dashboards.clj:
```diff
@@ -101,6 +101,10 @@
   [_entity-type segment-id-str]
   (api/read-check (db/select-one Segment :id (ensure-int segment-id-str))))
 
```
This enables the `/auto/dashboard/model/:id` endpoint (vs. question instead of model before).

markbastian committed:
Added dataset constraint to model api path

markbastian committed:
local changes only -- S&M this out later.

markbastian committed:
local changes only -- S&M this out later.

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Cleanup of M&D selection code

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
docs + test

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Tests of field-candidate-matching-test

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Testing field spec matching

markbastian commented on src/metabase/api/automagic_dashboards.clj:
```diff
@@ -101,6 +101,12 @@
   [_entity-type segment-id-str]
   (api/read-check (db/select-one Segment :id (ensure-int segment-id-str))))
 
```
Add the `/auto/dashboard/model/:id` route. Note that this will ensure the requested entity is a model, where `/auto/dashboard/question/:id` will not. The latter will still product a model x-ray for a model, but will do something else for questions.

markbastian commented on src/metabase/automagic_dashboards/core.clj:
```diff
@@ -243,11 +245,15 @@
 (defn- source
   [card]
   (cond
```
This is the fundamental logic change -- if the card is a model, x-ray it as a GenericTable. Nearly all other changes are documentation, tests, and refactors for better understanding.

markbastian committed:
linting

markbastian committed:
Another round of non-major dep upgrades

markbastian committed:
Revert hawk

markbastian committed:
Merge branch 'master' into update-misc-modules-29176

markbastian committed:
Reverting clojure.tools/namespace back to 1.4.0

markbastian committed:
hawk update

markbastian committed:
Some module updates

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Merge remote-tracking branch 'origin/x-ray-step-1-28674' into x-ray-step-1-28674

markbastian committed:
Removed comment that is probably not needed

dpsutton commented on src/metabase/automagic_dashboards/core.clj:
```diff
@@ -243,11 +245,15 @@
 (defn- source
   [card]
   (cond
+    ;; This is a non-native question with no ancestors (It is a query on a table).
```
I don't understand this comment. The only thing checked here is if the card is a model. Non-native, no ancestors, even "query on a table" can all be invalidated. I'm not seeing anything the comment really could say and this feels wrong.

markbastian replied:
Yeah, that might be historical and have gotten shifted around. Updating to simply "This is a model".

dpsutton commented on src/metabase/automagic_dashboards/core.clj:
```diff
@@ -378,39 +384,46 @@
                  (str/ends-with? name "_id"))))))
 
+(defn- field-spec-pred
+  "Generate a predicate of the form f(field) -> true | false based on a field spec."
```
`f(field)` is a weird way to write a function signature for a lisp

markbastian replied:
I was thinking "mathy" but I'll change to `(f field)`.

dpsutton commented on src/metabase/automagic_dashboards/core.clj:
```diff
@@ -378,39 +384,46 @@
                  (str/starts-with? name "id_")
                  (str/ends-with? name "_id"))))))
 
+(defn- field-spec-pred
```
what does `spec` mean in this matcher and the `named-spec-pred` and why is it absent from `max-cardinality-pred`?

It's also weird that the `field-spec-pred` matches either on the type or the name, and then there's another matcher to match on the regex.

Can these be broken out to `field-type-matcher`, `name-exact-matcher`, `name-regex-matcher`, `max-cardinality-matcher` or something like that?

markbastian replied:
Refactored to consistent naming and added more branch testing.

dpsutton commented on test/metabase/automagic_dashboards/core_test.clj:
```diff
@@ -208,6 +209,402 @@
             (test-automagic-analysis (db/select-one Card :id card-id) 8)))))))
 
+(deftest field-matching-predicates-test
+  (testing "The field spec predicate should match fields by their fieldspec"
```
This doesn't test the `(comp #{fieldspec} :name)` branch of `field-spec-pred`. (I actually think that's ok because it's weird that this matches on the semantic type OR the name if you give it a string)

markbastian replied:
Added a test. It's even weirder than that because the fieldspec in this branch has to be named "ga:fieldname."

dpsutton commented on test/metabase/automagic_dashboards/core_test.clj:
```diff
@@ -208,6 +209,402 @@
+        (is (true? ((#'magic/max-cardinality-pred 100) category-field))))))
+  (testing "Roll the above together and test filter-fields"
+    (mt/dataset sample-dataset
+      (let [category-field (db/select-one Field :id (mt/id :products :category))
```
this repetition is begging for a 
```clojure
  (defn field! [table column]
    (or (db/select-one Field :id (mt/id table column))
        (throw (ex-info (format "Did not find %s.%s" (name table) (name column))
                        {:table table :column column}))))
```

dpsutton commented on test/metabase/automagic_dashboards/core_test.clj:
```diff
@@ -208,6 +209,402 @@
+  (testing "A very simple card with two plain fields should return the singe assigned dimension for each field."
+    (mt/dataset sample-dataset
+      (mt/with-non-admin-groups-no-root-collection-perms
+        (let [source-query {:database (mt/id)
```
This query might be easier to read and write if you use `mt/$ids`

```clojure
(mt/$ids {:source-table $$orders
          :joins [{:fields [&u.people.state
                            &u.people.longitude
                            &u.people.latitude]
                   :source-table $$people
                   :condition [:= $orders.user_id &u.people.id]}]})
...
```

dpsutton reviewed (approved):
I'm approving but left a few comments that could be addressed for clarity.

markbastian committed:
Resolved 2 comments

markbastian committed:
version tweaks

markbastian committed:
more test coverage and better matcher names

markbastian committed:
locked shadow-cljs nrepl port to 50655

markbastian committed:
Merge branch 'master' into update-misc-modules-29176

markbastian committed:
Merge remote-tracking branch 'origin/update-misc-modules-29176' into x-ray-step-1-28674

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
updated several tests to use mt/$ids

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

ranquild committed:
Revert changes

ranquild committed:
Revert changes

ranquild committed:
Revert changes

ranquild committed:
Revert changes

ranquild committed:
Fix tests

ranquild committed:
Fix tests

ranquild committed:
Add a menu option to x-ray models

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

markbastian committed:
Merge branch '28472-questions-xray' into x-ray-step-1-28674

markbastian committed:
Merge remote-tracking branch 'origin/x-ray-step-1-28674' into x-ray-step-1-28674

# Conflicts:
#	frontend/src/metabase/components/EntityItem.jsx
#	frontend/src/metabase/lib/urls/questions.ts

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674

ranquild committed:
Fix the test

markbastian committed:
Merge branch 'master' into x-ray-step-1-28674
